api/napi/src/interpreter/component_instance.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-1.1 OR LicenseRef-Slint-commercial
+
+use napi::{Env, Error, JsFunction, JsUnknown, NapiRaw, NapiValue, Ref, Result};
+use slint_interpreter::{ComponentHandle, ComponentInstance, Value};
+
+use super::JsComponentDefinition;
+
+#[napi(js_name = "JsComponentInstance")]
+pub struct JsComponentInstance {
+    inner: ComponentInstance,
+}
+
+impl From<ComponentInstance> for JsComponentInstance {
+    fn from(instance: ComponentInstance) -> Self {
+        Self { inner: instance }
+    }
+}
+
+#[napi]
+impl JsComponentInstance {
+    #[napi(constructor)]
+    pub fn new() -> Self {
+        unreachable!("ComponentDefinition can only be created by using ComponentCompiler.")
+    }
+
+    #[napi]
+    pub fn definition(&self) -> JsComponentDefinition {
+        self.inner.definition().into()
+    }
+
+    #[napi]
+    pub fn run(&self) {
+        self.inner.run().unwrap()
+    }
+
+    #[napi]
+    pub fn get_property(&self, env: Env, name: String) -> Result<JsUnknown> {
+        let value = self
+            .inner
+            .get_property(name.as_ref())
+            .map_err(|e| Error::from_reason(format!("{e}")))?;
+        super::value::to_js_unknown(&env, &value)
+    }
+
+    #[napi]
+    pub fn set_property(&self, env: Env, name: String, js_value: JsUnknown) -> Result<()> {
+        let value = super::value::to_value(
+            &env,
+            &self.inner,
+            js_value,
+            &self
+                .inner
+                .get_property(&name)
+                .map_err(|_| napi::Error::from_reason("Cannot get property."))?,
```

Comment:
Setting a property should not end up reading it. You need to change the signature of `js_unknown_to_value` to take a `langtype::Type` as argument instead of an existing value, and then use the property _type_ here. The same applies to the return value of callbacks.

See also this code from the neon bindings that I mentioned on chat the other day, to determine the `langtype::Type` of a property:

```rust
 let ty = component.definition().properties_and_callbacks()
                .find_map(|(name, proptype)| if name == prop_name { Some(proptype) } else { None })
                .ok_or(())
                .or_else(|()| {
                    cx.throw_error(format!("Property {} not found in the component", prop_name))
                })?;
```